src/main/java/com/ironcorelabs/tenantsecurity/utils/CryptoUtils.java

```diff
@@ -0,0 +1,301 @@
+                                            .put(currentChunk).put(maybeTagBytes).array()));
+                        } else {
+                            decryptedStream.write(cipher.update(currentChunk));
+                            pushbackStream.unread(maybeTagBytes);
+                        }
+                    }
+                    return null;
+                });
+            });
+
+        });
+
+    }
+
+    public static Cipher getNewAesCipher(byte[] documentKey, byte[] iv, boolean encryptMode) throws Exception {
+        final Cipher cipher = Cipher.getInstance(AES_ALGO);
+        cipher.init(encryptMode ? Cipher.ENCRYPT_MODE : Cipher.DECRYPT_MODE, new SecretKeySpec(documentKey, "AES"),
+                new GCMParameterSpec(GCM_TAG_BIT_LENGTH, iv));
+        return cipher;
+    }
+
+    /**
+     * Given the provided document bytes and an AES key, encrypt and return the
+     * encrypted bytes.
+     */
+    public static CompletableFuture<byte[]> encryptBytes(byte[] document, DocumentMetadata metadata, byte[] documentKey,
+            SecureRandom secureRandom) {
+
+        // Create the output at a reasonable size. This means that the header can be up
+        // to 512 bytes of content without growing. This is a very comfortable upper
+        // bound for v3 header as only the tenant id is in there. If it needs to, the
+        // stream can grow, but this should help limit allocations.
+        ByteArrayOutputStream output = new ByteArrayOutputStream(
+                document.length + IV_BYTE_LENGTH + GCM_TAG_BYTE_LEN + HEADER_FIXED_SIZE_CONTENT_LENGTH + 512);
+        return encryptStreamInternal(documentKey, metadata, new ByteArrayInputStream(document), output, secureRandom)
+                .thenApply(unused -> {
+                    return output.toByteArray();
+                });
+    }
+
+    /**
+     * Decrypt a ICL document and return the plaintext.
+     *
+     * @param encryptedDocument The encrypted document including the ICL document
+     *                          header.
+     * @param documentKey       The AES key to decrypt the document.
+     * @return A future with the document decrypted.
+     */
+    public static CompletableFuture<byte[]> decryptDocument(byte[] encryptedDocumentBytes, byte[] documentKey) {
+        ByteArrayInputStream encryptedStream = new ByteArrayInputStream(encryptedDocumentBytes);
+        ByteArrayOutputStream decryptedStream = new ByteArrayOutputStream(encryptedDocumentBytes.length);
+        return decryptStreamInternal(documentKey, encryptedStream, decryptedStream)
+                .thenApply(unused -> decryptedStream.toByteArray());
+
+    }
+
+    /**
+     * Generate a header to mark the encrypted document as ours. Right now this is
+     * all constant; in the future this will contain a protobuf bytes header of
+     * variable length.
+     */
+    static CompletableFuture<byte[]> generateHeader(byte[] documentKey, DocumentMetadata metadata,
```

Comment:
Would it be appropriate here to insert a description of the full V3 header format? Maybe that's already specified somewhere else?

Reply:
I don't think it's needed here. The code for generating and parsing exists here and in the node TSC.